java/ql/src/experimental/Security/CWE/CWE-089/MyBatisMapperXmlSqlInjectionLib.qll

```diff
@@ -0,0 +1,124 @@
+      or
+      mbmxe.getInclude().getRefid() = mbms.getId() and
+      mbms.getAChild*() = xmle
+    ) and
+    mc.getMethod().getParameter(i).hasAnnotation() and
+    mc.getMethod().getParameter(i).getAnAnnotation() = annotation and
+    annotation.getType() instanceof TypeParam and
+    xmle.getTextValue()
+        .trim()
+        .matches("%${" + annotation.getValue("value").(CompileTimeConstantExpr).getStringValue() +
+            "%") and
+    mc.getArgument(i) = node.asExpr()
+  )
+  or
+  // MyBatis Mapper method Class Field sql injection vulnerabilities.
+  // e.g. MyBatis Mapper method: `void test(Test test);` and MyBatis Mapper XML file:`select id,name from test order by ${name,jdbcType=VARCHAR}`
+  exists(MyBatisMapperSqlOperation mbmxe, MyBatisMapperSql mbms, MethodAccess mc, int i, Class c |
+    mbmxe.getMapperMethod() = mc.getMethod()
+  |
+    (
+      mbmxe.getAChild*() = xmle
+      or
+      mbmxe.getInclude().getRefid() = mbms.getId() and
+      mbms.getAChild*() = xmle
+    ) and
+    not mc.getMethod().getParameter(i).hasAnnotation() and
+    mc.getMethod().getParameterType(i).getName() = c.getName() and
+    xmle.getTextValue().trim().matches("%${" + c.getAField().getName() + "%") and
+    mc.getArgument(i) = node.asExpr()
+  )
+  or
+  // The parameter type of MyBatis Mapper method is Map or List or Array, which may cause SQL injection vulnerability.
+  // e.g. MyBatis Mapper method: `void test(Map<String, String> params);` and MyBatis Mapper XML file:`select id,name from test where name like '%${name}%'`
+  exists(MyBatisMapperSqlOperation mbmxe, MyBatisMapperSql mbms, MethodAccess mc, int i |
+    mbmxe.getMapperMethod() = mc.getMethod()
+  |
+    (
+      mbmxe.getAChild*() = xmle
+      or
+      mbmxe.getInclude().getRefid() = mbms.getId() and
+      mbms.getAChild*() = xmle
+    ) and
+    not mc.getMethod().getParameter(i).hasAnnotation() and
+    (
+      mc.getMethod().getParameterType(i) instanceof MapType or
+      mc.getMethod().getParameterType(i) instanceof ListType or
+      mc.getMethod().getParameterType(i) instanceof Array
+    ) and
+    xmle.getTextValue().trim().matches("%${%") and
+    mc.getArgument(i) = node.asExpr()
+  )
+  or
+  // MyBatis Mapper method string type sql injection vulnerabilities.
+  // e.g. MyBatis Mapper method: `void test(String name);` and MyBatis Mapper XML file:`select id,name from test where name like '%${value}%'`
+  exists(MyBatisMapperSqlOperation mbmxe, MyBatisMapperSql mbms, MethodAccess mc |
+    mbmxe.getMapperMethod() = mc.getMethod()
+  |
+    (
+      mbmxe.getAChild*() = xmle
+      or
+      mbmxe.getInclude().getRefid() = mbms.getId() and
+      mbms.getAChild*() = xmle
+    ) and
+    mc.getMethod().getAParamType() instanceof TypeString and
+    mc.getMethod().getNumberOfParameters() = 1 and
+    not mc.getMethod().getAParameter().hasAnnotation() and
+    xmle.getTextValue().trim().matches("%${%") and
+    mc.getAnArgument() = node.asExpr()
+  )
+}
```

Comment:
@haby0 This clause is causing some false positives such as:

```java
@CheckForNull
UserDto selectUserByLogin(String login);
```

and

```xml
    <select id="selectUserByLogin" parameterType="string" resultType="User">
        SELECT
        <include refid="userColumns"/>
        FROM users u
        WHERE u.login=#{id, jdbcType=INTEGER} AND u.active=${_true}
    </select>
``` 

In this case, `_true` comes from:

```
  MyBatisConfBuilder(Database database) {
    this.conf = new Configuration();
    this.conf.setEnvironment(new Environment("production", createTransactionFactory(), database.getDataSource()));
    this.conf.setUseGeneratedKeys(true);
    this.conf.setLazyLoadingEnabled(false);
    this.conf.setJdbcTypeForNull(JdbcType.NULL);
    Dialect dialect = database.getDialect();
    this.conf.setDatabaseId(dialect.getId());
    this.conf.getVariables().setProperty("_true", dialect.getTrueSqlValue());
    this.conf.getVariables().setProperty("_false", dialect.getFalseSqlValue());
    this.conf.getVariables().setProperty("_from_dual", dialect.getSqlFromDual());
    this.conf.getVariables().setProperty("_scrollFetchSize", String.valueOf(dialect.getScrollDefaultFetchSize()));
    this.conf.setLocalCacheScope(LocalCacheScope.STATEMENT);
  }
```

Is the `login` java method parameter mapped to the `id` xml variable? If so, I guess that users can use arbitrary names and we cannot check for `${param_name%`. If thats the case, we may want to, at least, discard matches where the JDBC type is numeric

Reply:
Yes, `login` is mapped to `id`. If it is `$(id, jdbcType=INTEGER)`, there is also SQL injection.

Reply:
The false positive here is `${_true}`

Reply:
https://github.com/github/codeql/pull/6319/files#diff-ea9d95f3b6a6414f6cc9b69c5e4aba8ec7004dfb578048735565ae5b56257b17R29-R33

I think this sanitizer is ok.

Reply:
@pwntester I have eliminated the false positives caused by Mybatis Configuration Variable, please review it first.

Reply:
Can you check your new query against SonarQube project (eg: https://lgtm.com/query/7185819936285676051/). In the previous version, you already had a similar sanitizer.

In this particular example, the data flows as a String variable and is only coerced into an Integer one by the `jdbcType` mapper annotation. So Im not sure, that sanitizer would be enough.

>  If it is $(id, jdbcType=INTEGER), there is also SQL injection.

Wouldnt the `INTEGER` jdbcType` coercion prevent exploitation of the issue?

Reply:
> Can you check your new query against SonarQube project (eg: https://lgtm.com/query/7185819936285676051/). In the previous version, you already had a similar sanitizer.

Here are the results of my test.
https://lgtm.com/query/4878079417308405433/

> Wouldnt the INTEGER jdbcType` coercion prevent exploitation of the issue?

Sorry, I forgot to send you the actual operating conditions of the environment. Please check the email.

Reply:
@haby0 havent received any emails, but please share these operating conditions here so we all benefit from them. thanks!

Reply:
The attachment has been sent to your mailbox.

Reply:
Thanks, for the record, it seems that the type coercion is not enforced. Probably only applied when the value is NULL
@haby will wait your response about the `@Select` annotation before starting a new LGTM run

Reply:
understand.